yjit/src/cruby.rs

```diff
@@ -544,6 +545,75 @@ impl VALUE {
         let k: usize = item.wrapping_add(item.wrapping_add(1));
         VALUE(k)
     }
+
+    // Logic from internal/numeric.h rb_float_new_inline.
+    pub fn flonum_from_double(item: f64) -> Self {
```

Comment:
This seems much more involved than `fixnum_from_usize` to maintain a duplicated implementation. How about making a bindgen to `rb_float_new` (which is defined in `numeric.c`. In yjit.c, you could `#undef rb_float_value` and declare `double rb_float_value(VALUE v);` to reuse it from the bindgen) and call it?

Reply:
One thing is that ideally we probably want to avoid allocating on the Ruby heap while inside YJIT, just to reduce potential failure modes (eg triggering GC while YJIT is running).

I'd be OK with reusing the C implementation but ideally we would be able to encode to a flonum rather than allocating if possible.

Reply:
I mentioned in the meeting this week that I had it working with `rb_float_new`.
It was discussed that doing it without the C function would be best.
Maybe it would look less complicated if I removed some of the comments I added. 😄 
Or maybe feelings are different after seeing the algorithm. 😅 

The way CRuby does it is to encode these if they are within the MIN/MAX range, and anything outside of that needs to be on the heap.
It seems likely that YJIT won't need to use anything outside those bounds, so one option would be (like the `fixnum_from_usize` above) to just assert that we are within range.  Then we wouldn't need the C function and there would be no path for allocating.

If we don't want to duplicate this logic then yes we could use `rb_float_new` which has the same logic as this function: Encode values within range and create an object on the heap for anything else.

Reply:
> I'd be OK with reusing the C implementation but ideally we would be able to encode to a flonum rather than allocating if possible.
>
> I mentioned in the meeting this week that I had it working with rb_float_new.
> It was discussed that doing it without the C function would be best.

I'm not sure if I follow your logic. It seems like you assume that using the C implementation `rb_float_new` doesn't use flonum encoding, but it encodes a flonum whenever this Rust `flonum_from_double` would use the flonum encoding.

Similarly, it seems like you're saying this Rust implementation `flonum_from_double` doesn't allocate a Float object in the heap, but it does whenever `rb_float_new` would.

So, as long as you exactly replicate what the interpreter does, I don't think rewriting it in Rust changes anything with regard to "allocating on the Ruby heap" or "potential failure modes". If you're OK with reusing the C implementation, what is the motivation to duplicate the logic?

Reply:
I did not look at the implementation of `rb_float_new`. I assumed that it allocated because of `_new` in the name. If it does use flonum encoding when possible, then I agree it makes sense to use the C function. It's very unlikely a float will ever be allocated.

Reply:
This rust function is a direct port of `rb_float_new`: it does the transmute and encodes it if it's within range.  Out of range it falls back to `rb_float_new_in_heap`.

https://github.com/ruby/ruby/blob/002e785e7756e5105910434586a5752c8772bb2f/internal/numeric.h#L244-L273

When I saw the other stuff in this file about "not linking against CRuby in tests" I thought that was another reason we wanted it this way.

Is there any value in dropping the `rb_float_new_in_heap` call from here and just asserting that any floats yjit is using are in range?
Or is it best to just use `rb_float_new` and not duplicate any logic?

Reply:
> When I saw the other stuff in this file about "not linking against CRuby in tests" I thought that was another reason we wanted it this way.

We can't test Ruby's Float values in `cargo test`, so you'd need to test the result in `make test-all` anyway (if you want). So "not linking against CRuby in tests" doesn't help in that case. 

> Is there any value in dropping the rb_float_new_in_heap call from here and just asserting that any floats yjit is using are in range?

Leaving a `panic` for `#[cfg(test)]` is fine, but I don't think it's a good idea to potentially crash `RubyVM::YJIT.runtime_stats`. Remember, for the stats implementation, we are already allocating a Hash object using a C function. Nothing really changes even if you make sure there's no GC or C call for making a Float value. I still don't get what you're trying to solve with it.

> Or is it best to just use rb_float_new and not duplicate any logic?

I think so.

Reply:
I removed the implementation and linked to `rb_float_new` which simplifies a lot.

Regarding

> In yjit.c, you could #undef rb_float_value and declare double rb_float_value(VALUE v); to reuse it from the bindgen

I didn't seem to need this, if you think it's valuable, can you explain that a bit more?

Reply:
:+1:

Since bindgen finds `rb_float_new` successfully, it seems like you don't need to undef it.